controller/src/main/java/io/pravega/controller/server/bucket/UniformBucketDistributor.java

```diff
@@ -0,0 +1,141 @@
+ *     -- Add this to the controller containing min number of buckets.
+ * - The balanced condition happens when the difference between min and max buckets is at most 1.
+ */
+@Slf4j
+public class UniformBucketDistributor implements BucketDistributor {
+    @Override
+    public Map<String, Set<Integer>> distribute(Map<String, Set<Integer>> previousBucketControllerMapping,
+                                                @NotNull Set<String> currentControllers, int bucketCount) {
+        Preconditions.checkNotNull(previousBucketControllerMapping, "previousBucketControllerMapping");
+        Preconditions.checkNotNull(currentControllers, "currentControllers");
+        if (previousBucketControllerMapping.keySet().equals(currentControllers)) {
+            //Assuming the input map is always uniform distributed , since this distributor only depends on the controllers list.
+            log.debug("No change in controller list, using existing map {}", previousBucketControllerMapping);
+            return new HashMap<>(previousBucketControllerMapping);
+        }
+
+        if (previousBucketControllerMapping.size() == 0 || currentControllers.size() == 1) {
+            log.info("Creating new balanced map");
+            return initializeMap(currentControllers, bucketCount);
+        }
+
+        //Detect controllers not present in the existing Map.
+        Set<String> controllersAdded = currentControllers.stream().
+                                           filter(h -> !previousBucketControllerMapping.containsKey(h)).collect(Collectors.toSet());
+        if (!controllersAdded.isEmpty()) {
+            log.debug("New controllers added: {}", controllersAdded);
+        }
+        //Detect all buckets whose controllers has been removed. These need to be assigned to new controllers.
+        Set<Integer> orphanedBuckets = previousBucketControllerMapping.entrySet().stream().
+                                                             filter(h -> !currentControllers.contains(h.getKey()))
+                                                                   .flatMap(p -> p.getValue().stream())
+                                                                   .collect(Collectors.toSet());
+
+        //Using a TreeSet sorted by number of buckets to optimize add/remove operations.
+        TreeSet<Map.Entry<String, Set<Integer>>> mapElements =
+                new TreeSet<>((o1, o2) -> {
+                    if (o1.getValue().size() < o2.getValue().size()) {
+                        return -1;
+                    } else if (o1.getValue().size() > o2.getValue().size()) {
+                        return 1;
+                    } else {
+                        // Position elements with equal bucket length using object hashCode to provide strict weak
+                        // ordering. We don't need any specific ordering using information from the controller.
+                        return o1.getKey().hashCode() - o2.getKey().hashCode();
```

Comment:
Here we calculating the hash code of key which is of string type. Only leader will be responsible for distributing the buckets. As per my understanding hash code will remains same as it will get calculated on leader only.

Reply:
Perhaps `HashHelper.hash(String str)` would help.